### PR TITLE
Change start idle period algorithm to avoid monkey patching event loop.

### PR DESCRIPTION
**Pull request by rmcilroy**
Addresses #57

### CODE REVIEW

**Comment by rmcilroy** [comment]
@annevk, @igrigorik, @toddreifsteck please take a look and let me know what you think.

**Comment by igrigorik** [comment]
Preview @ https://cdn.rawgit.com/rmcilroy/requestidlecallback/event_loop/index.html

Ross, don't think it's due to this PR, but I see a couple of broken links in the preview:
- [current high resolution time] 
- [Spin the event loop]

File: index.html [hunk 1]
```diff
@@ -325,24 +325,29 @@
...
+        <li>Increment the <var>window</var>'s <a>idle callback identifier</a>
+        by one.</li>
+        <li>Let <var>handle</var> be the current value of <var>window</var>'s
+        <a>idle callback identifier</a>.</li>
+        <li>If the <var>window</var>'s <a>list of idle request callbacks</a> and
+        it's <a>list of runnable idle callbacks</a> are empty, [queue a task] on
+        the queue associated with the idle-task [task source], which performs
+        the <a>start an idle period algorithm</a>, passing <var>window</var> as
+        a parameter.</li>
+        <li>Push <var>callback</var> to the end of <var>window</var>'s <a>list
```
Comment (igrigorik) [inline index.html]: Should the push step precede the queue task? Any races here?

File: index.html [hunk 2]
```diff
@@ -376,35 +381,39 @@
...
         </li>
       </ol>
-      <p class="note">`requestIdleCallback` only schedules a single callback,
-      which will be executed during a single <a>idle period</a>. If the
-      callback cannot complete its work before the given <var>deadline</var>
-      then it should call `requestIdleCallback` again (which may be done from
-      within the callback) to schedule a future callback for the continuation
-      of its task, and exit immediately to return control back to the [event
-      loop].</p>
+      <p class="note"><code><a data-link-for="Window">requestIdleCallback</a>
```
Comment (igrigorik) [inline index.html]: fwiw, I think you can pull the data-link-for's up into the `<p data-link..>` (or even higher in the tree) and just `<a>requestIdleCallback</a>`.

File: index.html [hunk 3]
```diff
@@ -436,113 +445,116 @@
...
+          one idle period every 10 seconds to optimize for power usage.</p>
+        </li>
         <li>Let <var>now</var> be the current time.</li>
         <li>Let <var>deadline</var> be a time in the future until which the
-        browser expects to remain idle.</li>
+        browser expects to remain idle.
+          <p class='note'>The expectation is that the user agent will choose
+          <var>deadline</var> to ensure that no time-critical tasks will be
+          delayed even if a callback runs for the whole time period from
+          <var>now</var> to <var>deadline</var>. As such, it should be set to the
```
Comment (igrigorik) [inline index.html]: This reads like normative text (should ...), we should pull it out of a note.

File: index.html [hunk 4]
```diff
@@ -436,113 +445,116 @@
...
+          minimum of: the closest timeout in the [list of active timers] as set
+          via `[setTimeout]` and `[setInterval]`; the scheduled runtime for
+          pending animation callbacks posted via `[requestAnimationFrame]`;
+          pending internal timeouts such as deadlines to start rendering the
+          next frame, process audio or any other internal task the user agent
+          deems important; and a maximum cap of 50ms in the future to ensure
+          responsiveness to unpredictable user input within the threshold of
```
Comment (igrigorik) [inline index.html]: nitpick: s/unpredictable/new ?

File: index.html [hunk 5]
```diff
@@ -436,113 +445,116 @@
...
             <var>deadline</var> and set the `didTimeout` attribute to
             `false`.</li>
             <li>Call <var>callback</var> with <var>deadlineArg</var> as its
             argument. If an uncaught runtime script error occurs, then [report
             the error].</li>
-            <li>If <var>docs</var> is not empty, the user agent SHOULD [queue a
-            task] which performs the steps in the <a>invoke idle callbacks
-            algorithm</a> with <var>deadline</var> as parameter.</li>
+            <li>If <var>window</var>'s <a>list of runnable idle callbacks</a>
+            is not empty, the user agent SHOULD [queue a task] which performs
```
Comment (igrigorik) [inline index.html]: should the SHOULD be dropped here? What happens if the UA skips this step?

File: index.html [hunk 6]
```diff
@@ -436,113 +445,116 @@
...
+            the steps in the <a>invoke idle callbacks algorithm</a> with
+            <var>deadline</var> and <var>window</var> as a parameters and return
+            from this algorithm</li>
           </ol>
         </li>
+        <li>Otherwise, if either the <var>window</var>'s <a>list of idle request
+        callbacks</a> or it's <a>list of runnable idle callbacks</a> are
+        not empty, the user agent SHOULD [queue a task] which performs the steps
```
Comment (igrigorik) [inline index.html]: Same as above, we want to make sure idle callbacks are not dropped?

**Review by igrigorik — COMMENTED** [review]
A few comments and questions. 

I'll defer to @annevk @domenic for the task + event loop plumbing.

**Comment by annevk** [comment]
If there are no tasks on the event loop, you won't necessarily have an animation frame first.

**Comment by rmcilroy** [comment]
re https://github.com/w3c/requestidlecallback/pull/60#issuecomment-318992607, I'm not sure why this would be an issue. If there are no other tasks / animation frames then we should keep cycling idle periods every 50ms while there are still idle callbacks. Once new tasks / animation frames are posted we should drop back to enabling idle periods in the time between the end of rendering an animation frame and the expected time that the next frame will start running task. Is there something I'm missing?

**Comment by annevk** [comment]
I thought the requirement was that it happened after the animation frame.

**Comment by rmcilroy** [comment]
The requirement is that if there is a pending animation frame the idle period must happen after that frame is rendered, otherwise if no frames are being drawn the the idle periods can start any time there are no other tasks in the event loop.

It's difficult to express this in relation to the HTML event loop spec though since there isn't really a concept of "pending animation frame", instead the event loop spec is written such that each spin of the even loop the user agent renders updates for all documents, except those documents that the "user agent believes would not benefit from having its rendering updated at this time". From the attached note it is clear this is intended to enable the browser to only have to render once per frame, but it's not exactly how it's speced.

**Comment by annevk** [comment]
Hmm, I wonder if we should change that around. Ugh.

I guess this is probably okay then, though I'm a little wary of this depending on a bunch of specific things without reference back from HTML. Seems easy for this to go out of sync.

**Comment by igrigorik** [comment]
@rmcilroy any thoughts on my review comments above? If those are no-ops, it sounds like we can merge this.

**Comment by rmcilroy** [comment]
Thanks for the comments @annevk. I'll see if I can include some more specific references back to HTML where it makes sense to reduce the scope for this going out of sync.

@igrigorik yeah I think some of your comments definitely make sense, I just wanted to ensure the overall algorithm was workable before addressing them, I'll take a look at them now.

File: index.html (hunk 1 above)
Comment (rmcilroy) [reply index.html]: Done.

File: index.html (hunk 2 above)
Comment (rmcilroy) [reply index.html]: Done.

File: index.html (hunk 3 above)
Comment (rmcilroy) [reply index.html]: Done.

File: index.html (hunk 4 above)
Comment (rmcilroy) [reply index.html]: Done.

File: index.html (hunk 5 above)
Comment (rmcilroy) [reply index.html]: Right this was originally to give scope for ending an idle period early, but that is now done in step one. Done.

File: index.html (hunk 6 above)
Comment (rmcilroy) [reply index.html]: Likewise, done.

**Comment by rmcilroy** [comment]
> Ross, don't think it's due to this PR, but I see a couple of broken links in the preview:

Fixed.

Comments addressed, PTAL thanks.

**Comment by igrigorik** [comment]
Looks good! I see one broken reference, however:

image

**Comment by rmcilroy** [comment]
Dang, thought I'd fixed that link. Should be fixed now, PTAL thanks.

**Review by igrigorik — APPROVED** [review]
Looks great, thanks Ross!

**Comment by igrigorik** [comment]
Merging. 

@annevk @toddreifsteck if you have any other feedback, please ping here or open a new issue.

**Pull request merged**